[mypy] Incorrect type for argument `data` to `httpx.AsyncClient.request`

Incorrect annotation for argument `data` to `httpx.AsyncClient.request` as referenced in https://github.com/brettcannon/gidgethub/issues/74#issuecomment-723391219

Location in the module: `gidgethub/httpx.py::GitHubAPI::_request(body: bytes = b"")`

Reference:
`httpx` client module: https://github.com/encode/httpx/blob/master/httpx/_client.py#L1334
Type source: https://github.com/encode/httpx/blob/master/httpx/_types.py#L73

```console
gidgethub/httpx.py:19: error: Argument "data" to "request" of "AsyncClient" has incompatible type "bytes"; expected "Dict[Any, Any]"
```

--- a/docs/abc.rst
+++ b/docs/abc.rst
@@ -195,7 +195,7 @@ experimental APIs without issue.
         For this situation, you can pass ``data=b""``.
 
 
-        .. versionchanged:: 4.12
+        .. versionchanged:: 4.1.2
             Added *content_type*.
 
 

--- a/docs/abc.rst
+++ b/docs/abc.rst
@@ -195,7 +195,7 @@ experimental APIs without issue.
         For this situation, you can pass ``data=b""``.
 
 
-        .. versionchanged:: 4.1.2
+        .. versionchanged:: 4.2.0
             Added *content_type*.
 
 

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -7,6 +7,9 @@ Changelog
 - Add :meth:`gidgethub.routing.Router.fetch` for obtaining a frozenset of functions
   registered to the router that the event would be called on.
   (`Issue #74 <https://github.com/brettcannon/gidgethub/issues/74>`_).
+- Make router callback execution order non-deterministic to avoid relying on
+   registration order.
+   (`Issue #74 <https://github.com/brettcannon/gidgethub/issues/74>`_).
 - Fix mypy errors in ``gidgethub.httpx.GitHubAPI._request``
   (`Issue #133 <https://github.com/brettcannon/gidgethub/issues/133>`_).
 

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -9,6 +9,11 @@ Changelog
   and the raw form of the data will be passed to the request. If not supplied,
   by default the data will be parsed as JSON, and the "application/json" content
   type will be used. (`Issue #115 <https://github.com/brettcannon/gidgethub/issues/115>`_).
+- Add :meth:`gidgethub.routing.Router.fetch` for obtaining a frozenset of functions 
+  registered to the router that the event would be called on.
+  (`Issue #74 <https://github.com/brettcannon/gidgethub/issues/74>`_).
+- Fix mypy errors in ``gidgethub.httpx.GitHubAPI._request`` 
+  (`Issue #133 <https://github.com/brettcannon/gidgethub/issues/133>`_).
 
 
 4.1.1

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -1,6 +1,15 @@
 Changelog
 =========
 
+5.0.0
+-----
+
+- Add :meth:`gidgethub.routing.Router.fetch` for obtaining a frozenset of functions
+  registered to the router that the event would be called on.
+  (`Issue #74 <https://github.com/brettcannon/gidgethub/issues/74>`_).
+- Fix mypy errors in ``gidgethub.httpx.GitHubAPI._request``
+  (`Issue #133 <https://github.com/brettcannon/gidgethub/issues/133>`_).
+
 4.2.0
 -----
 
@@ -9,12 +18,6 @@ Changelog
   and the raw form of the data will be passed to the request. If not supplied,
   by default the data will be parsed as JSON, and the "application/json" content
   type will be used. (`Issue #115 <https://github.com/brettcannon/gidgethub/issues/115>`_).
-- Add :meth:`gidgethub.routing.Router.fetch` for obtaining a frozenset of functions 
-  registered to the router that the event would be called on.
-  (`Issue #74 <https://github.com/brettcannon/gidgethub/issues/74>`_).
-- Fix mypy errors in ``gidgethub.httpx.GitHubAPI._request`` 
-  (`Issue #133 <https://github.com/brettcannon/gidgethub/issues/133>`_).
-
 
 4.1.1
 -----

--- a/docs/routing.rst
+++ b/docs/routing.rst
@@ -79,6 +79,8 @@ in user code.
         to the router that the *event* would be called on. The *event*
         argument corresponds to :class:`gidgethub.sansio.Event`.
 
+        .. versionadded:: 4.1.2
+
 
     .. coroutine:: dispatch(event, *args, **kwargs)
 

--- a/docs/routing.rst
+++ b/docs/routing.rst
@@ -79,7 +79,7 @@ in user code.
         to the router that the *event* would be called on. The *event*
         argument corresponds to :class:`gidgethub.sansio.Event`.
 
-        .. versionadded:: 4.2.0
+        .. versionadded:: 5.0.0
 
 
     .. coroutine:: dispatch(event, *args, **kwargs)

--- a/docs/routing.rst
+++ b/docs/routing.rst
@@ -73,6 +73,13 @@ in user code.
                 ...
 
 
+    .. method:: fetch(event)
+
+        Return a set of asynchronous callbacks registered
+        to the router that the *event* would be called on. The *event*
+        argument corresponds to :class:`gidgethub.sansio.Event`.
+
+
     .. coroutine:: dispatch(event, *args, **kwargs)
 
         Call the appropriate asynchronous callbacks for the *event*.

--- a/docs/routing.rst
+++ b/docs/routing.rst
@@ -90,3 +90,6 @@ in user code.
 
         .. versionchanged:: 2.4
             Added ``*args`` and ``**kwargs``.
+            
+        .. versionchanged:: 5.0.0
+            Execution order is non-deterministic.

--- a/docs/routing.rst
+++ b/docs/routing.rst
@@ -79,7 +79,7 @@ in user code.
         to the router that the *event* would be called on. The *event*
         argument corresponds to :class:`gidgethub.sansio.Event`.
 
-        .. versionadded:: 4.1.2
+        .. versionadded:: 4.2.0
 
 
     .. coroutine:: dispatch(event, *args, **kwargs)

--- a/docs/routing.rst
+++ b/docs/routing.rst
@@ -75,7 +75,7 @@ in user code.
 
     .. method:: fetch(event)
 
-        Return a set of asynchronous callbacks registered
+        Return a frozenset of asynchronous callbacks registered
         to the router that the *event* would be called on. The *event*
         argument corresponds to :class:`gidgethub.sansio.Event`.
 

--- a/gidgethub/httpx.py
+++ b/gidgethub/httpx.py
@@ -16,7 +16,7 @@ class GitHubAPI(gh_abc.GitHubAPI):
     ) -> Tuple[int, Mapping[str, str], bytes]:
         """Make an HTTP request."""
         response = await self._client.request(
-            method, url, headers=dict(headers), data=body
+            method, url, headers=dict(headers), content=body
         )
         return response.status_code, response.headers, response.content
 

--- a/gidgethub/routing.py
+++ b/gidgethub/routing.py
@@ -1,4 +1,4 @@
-from typing import Any, Awaitable, Callable, Dict, List, Optional, Set
+from typing import Any, Awaitable, Callable, Dict, List, Set
 
 from . import sansio
 
@@ -35,7 +35,6 @@ class Router:
         key of the data in the event being dispatched.
         """
         if len(data_detail) > 1:
-            msg = ()
             raise TypeError(
                 "dispatching based on data details is only "
                 "supported up to one level deep; "

--- a/gidgethub/routing.py
+++ b/gidgethub/routing.py
@@ -1,4 +1,4 @@
-from typing import Any, Awaitable, Callable, Dict, List, Optional
+from typing import Any, Awaitable, Callable, Dict, List, Optional, Set
 
 from . import sansio
 
@@ -62,12 +62,12 @@ class Router:
 
         return decorator
 
-    async def dispatch(self, event: sansio.Event, *args: Any, **kwargs: Any) -> None:
-        """Dispatch an event to all registered function(s)."""
-
-        found_callbacks = []
+    def fetch(self, event: sansio.Event) -> Set[AsyncCallback]:
+        """Return a set of function(s) registered to the router that the event would
+        be called on."""
+        found_callbacks = set()
         try:
-            found_callbacks.extend(self._shallow_routes[event.event])
+            found_callbacks.update(self._shallow_routes[event.event])
         except KeyError:
             pass
         try:
@@ -79,6 +79,12 @@ class Router:
                 if data_key in event.data:
                     event_value = event.data[data_key]
                     if event_value in data_values:
-                        found_callbacks.extend(data_values[event_value])
+                        found_callbacks.update(data_values[event_value])
+        return found_callbacks
+
+    async def dispatch(self, event: sansio.Event, *args: Any, **kwargs: Any) -> None:
+        """Dispatch an event to all registered function(s)."""
+
+        found_callbacks = self.fetch(event)
         for callback in found_callbacks:
             await callback(event, *args, **kwargs)

--- a/gidgethub/routing.py
+++ b/gidgethub/routing.py
@@ -1,4 +1,4 @@
-from typing import Any, Awaitable, Callable, Dict, List, Set
+from typing import Any, Awaitable, Callable, Dict, List, FrozenSet
 
 from . import sansio
 
@@ -61,7 +61,7 @@ class Router:
 
         return decorator
 
-    def fetch(self, event: sansio.Event) -> Set[AsyncCallback]:
+    def fetch(self, event: sansio.Event) -> FrozenSet[AsyncCallback]:
         """Return a set of function(s) registered to the router that the event would
         be called on."""
         found_callbacks = set()
@@ -79,7 +79,7 @@ class Router:
                     event_value = event.data[data_key]
                     if event_value in data_values:
                         found_callbacks.update(data_values[event_value])
-        return found_callbacks
+        return frozenset(found_callbacks)
 
     async def dispatch(self, event: sansio.Event, *args: Any, **kwargs: Any) -> None:
         """Dispatch an event to all registered function(s)."""

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -32,7 +32,7 @@ dev = ["aiohttp", "black", "coverage[toml]>=5.0.3", "httpx", "mypy", "pytest-cov
        "pytest-xdist", "tornado"]
 aiohttp = ["aiohttp"]
 tornado = ["tornado"]
-httpx = ["httpx>=0.11.0"]
+httpx = ["httpx>=0.16.1"]
 
 [tool.flit.metadata.urls]
 Documentation = "https://gidgethub.readthedocs.io"

--- a/tests/test_routing.py
+++ b/tests/test_routing.py
@@ -152,7 +152,8 @@ def test_fetch_callbacks():
     # Wrong data detail value.
     router.add(never_called.meth, "yeah", count=-13)
     no_callbacks = router.fetch(event)
-    assert not no_callbacks  # no callbacks found.
+    assert not no_callbacks  # No callbacks found.
+    assert not never_called.called
 
     shallow_registration_1 = Callback()
     shallow_registration_2 = Callback()
@@ -168,3 +169,7 @@ def test_fetch_callbacks():
     assert shallow_registration_2.meth not in callbacks
     assert deep_registration_1.meth not in callbacks
     assert deep_registration_2.meth in callbacks
+    assert not shallow_registration_1.called
+    assert not shallow_registration_2.called
+    assert not deep_registration_1.called
+    assert not deep_registration_2.called

--- a/tests/test_routing.py
+++ b/tests/test_routing.py
@@ -140,7 +140,7 @@ def test_fetch_callbacks():
         delivery_id="1234",
     )
     no_callbacks = router.fetch(event)
-    assert not no_callbacks  # No callbacks registered
+    assert not no_callbacks  # No callbacks registered.
 
     never_called = Callback()
     # Wrong event.
@@ -152,7 +152,7 @@ def test_fetch_callbacks():
     # Wrong data detail value.
     router.add(never_called.meth, "yeah", count=-13)
     no_callbacks = router.fetch(event)
-    assert not no_callbacks  # no callbacks found
+    assert not no_callbacks  # no callbacks found.
 
     shallow_registration_1 = Callback()
     shallow_registration_2 = Callback()

--- a/tests/test_routing.py
+++ b/tests/test_routing.py
@@ -130,3 +130,41 @@ async def test_router_copy():
     await other_router.dispatch(event)
     assert deep_callback.called
     assert shallow_callback.called
+
+
+def test_fetch_callbacks():
+    router = routing.Router()
+    event = sansio.Event(
+        {"action": "opened", "count": 42, "status": "completed"},
+        event="yeah",
+        delivery_id="1234",
+    )
+    no_callbacks = router.fetch(event)
+    assert not no_callbacks  # No callbacks registered
+
+    never_called = Callback()
+    # Wrong event.
+    router.add(never_called.meth, "nope")
+    # Wrong event and data detail.
+    router.add(never_called.meth, "nope", action="new")
+    # Wrong data detail key.
+    router.add(never_called.meth, "yeah", never=42)
+    # Wrong data detail value.
+    router.add(never_called.meth, "yeah", count=-13)
+    no_callbacks = router.fetch(event)
+    assert not no_callbacks  # no callbacks found
+
+    shallow_registration_1 = Callback()
+    shallow_registration_2 = Callback()
+    deep_registration_1 = Callback()
+    deep_registration_2 = Callback()
+    router = routing.Router()
+    router.add(shallow_registration_1.meth, "yeah")
+    router.add(shallow_registration_2.meth, "nope")
+    router.add(deep_registration_1.meth, "nope", status="completed")
+    router.add(deep_registration_2.meth, "yeah", count=42)
+    callbacks = router.fetch(event)
+    assert shallow_registration_1.meth in callbacks
+    assert shallow_registration_2.meth not in callbacks
+    assert deep_registration_1.meth not in callbacks
+    assert deep_registration_2.meth in callbacks

--- a/tests/test_routing.py
+++ b/tests/test_routing.py
@@ -151,8 +151,8 @@ def test_fetch_callbacks():
     router.add(never_called.meth, "yeah", never=42)
     # Wrong data detail value.
     router.add(never_called.meth, "yeah", count=-13)
-    no_callbacks = router.fetch(event)
-    assert not no_callbacks  # No callbacks found.
+    callbacks = router.fetch(event)
+    assert not callbacks  # No callbacks found.
     assert not never_called.called
 
     shallow_registration_1 = Callback()

--- a/tests/test_routing.py
+++ b/tests/test_routing.py
@@ -139,8 +139,8 @@ def test_fetch_callbacks():
         event="yeah",
         delivery_id="1234",
     )
-    no_callbacks = router.fetch(event)
-    assert not no_callbacks  # No callbacks registered.
+    callbacks = router.fetch(event)
+    assert not callbacks  # No callbacks registered.
 
     never_called = Callback()
     # Wrong event.

